lib/maths/CBoostedTreeLoss.cc

```diff
@@ -512,10 +525,131 @@ CArgMinMsleImpl::TObjective CArgMinMsleImpl::objective() const {
+        m_Buckets[bucketIndex].add(error, weight);
+        break;
+    }
+    default:
+        break;
+    }
+}
+
+void CArgMinPseudoHuberImpl::merge(const CArgMinLossImpl& other) {
+    const auto* huber = dynamic_cast<const CArgMinPseudoHuberImpl*>(&other);
+    if (huber != nullptr) {
+        switch (m_CurrentPass) {
+        case 0:
+            m_ErrorMinMax += huber->m_ErrorMinMax;
+            break;
+        case 1:
+            for (std::size_t i = 0; i < m_Buckets.size(); ++i) {
+                m_Buckets[i] += huber->m_Buckets[i];
+            }
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+CArgMinPseudoHuberImpl::TDoubleVector CArgMinPseudoHuberImpl::value() const {
+    TObjective objective;
+    double minWeight = m_ErrorMinMax.min();
+    double maxWeight = m_ErrorMinMax.max();
+
+    objective = this->objective();
+    double minimizer;
+    double objectiveAtMinimum;
+    std::size_t maxIterations{HUBER_OPTIMIZATION_ITERATIONS};
+    CSolvers::minimize(minWeight, maxWeight, objective(minWeight), objective(maxWeight),
+                       objective, 1e-5, maxIterations, minimizer, objectiveAtMinimum);
+    LOG_TRACE(<< "minimum = " << minimizer << " objective(minimum) = " << objectiveAtMinimum);
+
+    TDoubleVector result(1);
+    result(0) = minimizer;
+    return result;
+}
+
+CArgMinPseudoHuberImpl::TObjective CArgMinPseudoHuberImpl::objective() const {
+    return [this](double weight) {
+        if (m_Delta2 > 0) {
+
+            double loss{0.0};
+            double totalCount{0.0};
+            for (const auto& bucket : m_Buckets) {
+                double count{CBasicStatistics::count(bucket)};
+                if (count > 0.0) {
+                    double error{CBasicStatistics::mean(bucket)};
+                    loss += count * m_Delta2 *
+                            (std::sqrt(1.0 + CTools::pow2(error - weight) / m_Delta2) - 1.0);
+                    totalCount += count;
+                }
+            }
+            return loss / totalCount + this->lambda() * CTools::pow2(weight);
+        } else {
+            return 0.0;
+        }
+    };
+}
 }
 
 namespace boosted_tree {
 
+CLoss::TLossUPtr CLoss::restoreLoss(core::CStateRestoreTraverser& traverser) {
+    do {
+        const std::string& lossFunctionName = traverser.name();
+        if (lossFunctionName == CMse::NAME) {
+            return std::make_unique<CMse>(traverser);
+        } else if (lossFunctionName == CMsle::NAME) {
+            return std::make_unique<CMsle>(traverser);
+        } else if (lossFunctionName == CPseudoHuber::NAME) {
+            return std::make_unique<CPseudoHuber>(traverser);
+        } else if (lossFunctionName == CBinomialLogisticLoss::NAME) {
+            return std::make_unique<CBinomialLogisticLoss>(traverser);
+        } else if (lossFunctionName == CMultinomialLogisticLoss::NAME) {
+            return std::make_unique<CMultinomialLogisticLoss>(traverser);
+        }
```

Comment:
This needs to be in a catch since the restore constructors can throw.

Reply:
Added a try-catch block